# **ACTIVIDAD 7 - Data Wrangling (Limpieza de datos)**

In [5]:
import pandas as pd
import os 

path=os.getcwd()
print(path)
file="Desaparecidos._Colombia__hist_rico__a_os_1930_a_marzo_de_2024_20240729.csv"
full_path = os.path.join(path,'DATA', file)
print(full_path)
tabla=pd.read_csv(full_path,sep=',',encoding='utf-8')
tabla.head()

c:\Users\angie.galicia\Downloads\python\BITrepository-main
c:\Users\angie.galicia\Downloads\python\BITrepository-main\DATA\Desaparecidos._Colombia__hist_rico__a_os_1930_a_marzo_de_2024_20240729.csv


ID Estado de la desaparición Clasificación de la desaparición  \
0   1              Desaparecido                  Sin Información   
1   2              Desaparecido                  Sin Información   
2   3              Desaparecido                  Sin Información   
3   4             Aparecio Vivo                  Sin Información   
4   5             Aparecio Vivo                  Sin Información   

  Fecha de la desaparición Edad al momento de la desaparición  \
0  2006-12-29T00:00:00.000                                 21   
1  2006-12-26T00:00:00.000                                 24   
2  2006-12-29T00:00:00.000                                 13   
3  2006-12-29T00:00:00.000                                 17   
4  2006-12-31T00:00:00.000                                 71   

  Sexo del desaparecido País de nacimiento del desaparecido  \
0                Hombre                     Sin información   
1                Hombre                            Colombia   
2                 Mujer                            Colombia   
3                 Mujer                     Sin información   
4                Hombre                            Colombia   

  Ancestro racial del desaparecido Estado civil del desaparecido  \
0                          Mestizo                   Soltero (a)   
1                          Mestizo                   Soltero (a)   
2                          Mestizo                   Soltero (a)   
3                          Mestizo                   Soltero (a)   
4                          Mestizo                   Unión Libre   

                    Escolaridad del desaparecido  \
0  Educación básica secundaria o secundaria baja   
1  Educación básica secundaria o secundaria baja   
2  Educación básica secundaria o secundaria baja   
3    Educación técnica profesional y tecnológica   
4                      Educación básica primaria   

              Pertenencia grupal del desaparecido  \
0  Persona adicta a una droga natural o sintética   
1  Persona adicta a una droga natural o sintética   
2                                         Ninguno   
3                                         Ninguno   
4                                            Otro   

  País donde ocurre la desaparición  \
0                          Colombia   
1                          Colombia   
2                          Colombia   
3                          Colombia   
4                          Colombia   

  Municipio donde ocurre la desaparición DANE  \
0                                 Bogotá D.C.   
1                                 Bogotá D.C.   
2                                 Bogotá D.C.   
3                                 Bogotá D.C.   
4                                 Bogotá D.C.   

  Departamento donde ocurre la desaparición DANE      Contexto  
0                                    Bogotá D.C.  Desaparición  
1                                    Bogotá D.C.  Desaparición  
2                                    Bogotá D.C.  Desaparición  
3                                    Bogotá D.C.  Desaparición  
4                                    Bogotá D.C.  Desaparición

**1. Ajustar nombre de columnas**

In [10]:
tabla.rename(columns={"Departamento donde ocurre la desaparición DANE": "Departamento"}, inplace=True)
tabla.rename(columns={"Estado de la desaparición":"Estado"}, inplace=True)
tabla.rename(columns={"Clasificación de la desaparición":"Clasificacion"}, inplace=True)
tabla.rename(columns={"Fecha de la desaparición":"Fecha_desaparición"}, inplace=True)
tabla.rename(columns={"Edad al momento de la desaparición":"Edad"}, inplace=True)
tabla.rename(columns={"Sexo del desaparecido":"Sexo"}, inplace=True)
tabla.rename(columns={"País de nacimiento del desaparecido":"Pais_nacimiento"}, inplace=True)
tabla.rename(columns={"Ancestro racial del desaparecido":"Raza"}, inplace=True)
tabla.rename(columns={"Estado civil del desaparecido":"Estado_civil"}, inplace=True)
tabla.rename(columns={"Escolaridad del desaparecido":"Escolaridad"}, inplace=True)
tabla.rename(columns={"Pertenencia grupal del desaparecido":"Pertenencia_grupal"}, inplace=True)
tabla.rename(columns={"País donde ocurre la desaparición":"Pais_desaparicion"}, inplace=True)
tabla.rename(columns={"Municipio donde ocurre la desaparición DANE":"Municipio"}, inplace=True)

tabla.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 185741 entries, 0 to 185740
Data columns (total 15 columns):
 #   Column              Non-Null Count   Dtype 
---  ------              --------------   ----- 
 0   ID                  185741 non-null  int64 
 1   Estado              185741 non-null  object
 2   Clasificacion       185741 non-null  object
 3   Fecha_desaparición  185741 non-null  object
 4   Edad                185741 non-null  object
 5   Sexo                185741 non-null  object
 6   Pais_nacimiento     185741 non-null  object
 7   Raza                185741 non-null  object
 8   Estado_civil        185741 non-null  object
 9   Escolaridad         185741 non-null  object
 10  Pertenencia_grupal  185741 non-null  object
 11  Pais_desaparicion   185741 non-null  object
 12  Municipio           185741 non-null  object
 13  Departamento        185741 non-null  object
 14  Contexto            185741 non-null  object
dtypes: int64(1), object(14)
memory usage: 21.3+ MB


**2. Validar duplicados**

In [11]:
duplicados = tabla.duplicated()
numero_duplicados = duplicados.sum()
print(f"Número de registros duplicados: {numero_duplicados}")

Número de registros duplicados: 0


**3. Validar registros Nulos**

In [13]:
nulos_por_columna = tabla.isnull().sum()
print(nulos_por_columna)

ID                    0
Estado                0
Clasificacion         0
Fecha_desaparición    0
Edad                  0
Sexo                  0
Pais_nacimiento       0
Raza                  0
Estado_civil          0
Escolaridad           0
Pertenencia_grupal    0
Pais_desaparicion     0
Municipio             0
Departamento          0
Contexto              0
dtype: int64


**4. Validar Outliers para el campo Edad**

In [ ]:
!pip install scikit-learn

In [ ]:
import numpy as np

# Reemplazar "Sin información" y "Menor de 1 año" por NaN para poder calcular los Outliers 
tabla['Edad'] = tabla['Edad'].replace({'Sin información': np.nan, 'Menor de 1 año': np.nan})

# Convertir la columna a tipo numérico, forzando errores a NaN
tabla['Edad'] = pd.to_numeric(tabla['Edad'], errors='coerce')

media_edad = tabla['Edad'].mean()
tabla['Edad'].fillna(media_edad, inplace=True)

In [54]:
# Calcular el IQR
Q1 = tabla['Edad'].quantile(0.25)
Q3 = tabla['Edad'].quantile(0.75)
IQR = Q3 - Q1

# Definir límites para identificar outliers
limite_inferior = Q1 - 1.5 * IQR
limite_superior = Q3 + 1.5 * IQR

# Filtrar outliers
print("Limite inferior Edad:", limite_inferior)
print("Limite superior Edad:", limite_superior)
outliers = tabla[(tabla['Edad'] < limite_inferior) | (tabla['Edad'] > limite_superior)]

# Eliminar outliers del DataFrame original
tabla_cleaned = tabla[~((tabla['Edad'] < limite_inferior) | (tabla['Edad'] > limite_superior))]

Limite inferior Edad: -14.5
Limite superior Edad: 69.5


In [27]:
print("Número de outliers:", len(outliers))
print("Tabla normal:", len(tabla))
print("Tabla quitando Outliers:", len(tabla_cleaned))

Número de outliers: 5490
Tabla normal: 185741
Tabla quitando Outliers: 180251


**5. Como en el dataset no existen datos Nulos, si existen datos con el dato "Sin informacion" por lo que se busca los registros con mayor cantidad de columnas con este dato para limpiar**

In [59]:
# Contar cuántas columnas tienen el valor "Sin información"
conteo_sin_info = tabla_cleaned.apply(lambda x: (x == "Sin Información").sum(), axis=1)
conteo_sin_info.max()

np.int64(5)

In [60]:
umbral = 4

# Filtrar registros
registros_filtrados = tabla_cleaned[conteo_sin_info > umbral]
print("Registros con mayoría de 'Sin información':")
registros_filtrados

Registros con mayoría de 'Sin información':


ID        Estado    Clasificacion       Fecha_desaparición  Edad  \
1872      1873  Desaparecido  Sin Información  2006-06-01T00:00:00.000  30.0   
2789      2790  Desaparecido  Sin Información  2003-06-28T00:00:00.000  50.0   
3750      3751  Desaparecido  Sin Información  2006-09-01T00:00:00.000  50.0   
3818      3819  Desaparecido  Sin Información  2007-01-01T00:00:00.000  50.0   
4310      4311  Desaparecido  Sin Información  2007-01-01T00:00:00.000  19.0   
...        ...           ...              ...                      ...   ...   
151555  151743  Desaparecido  Sin Información  2005-09-20T00:00:00.000  30.0   
151740  151928  Desaparecido  Sin Información  2003-06-29T00:00:00.000  29.0   
174863  175091  Desaparecido  Sin Información  2008-06-09T00:00:00.000  20.0   
175796  176024  Desaparecido  Sin Información  2002-02-01T00:00:00.000  42.0   
184542  184793  Desaparecido  Sin Información  2015-06-04T00:00:00.000  35.0   

          Sexo  Pais_nacimiento             Raza     Estado_civil  \
1872    Hombre  Sin información  Sin Información  Sin Información   
2789    Hombre  Sin información  Sin Información  Sin Información   
3750    Hombre  Sin información  Sin Información  Sin Información   
3818    Hombre  Sin información  Sin Información  Sin Información   
4310    Hombre         Colombia  Sin Información  Sin Información   
...        ...              ...              ...              ...   
151555  Hombre         Colombia  Sin Información  Sin Información   
151740  Hombre  Sin información  Sin Información  Sin Información   
174863  Hombre  Sin información  Sin Información  Sin Información   
175796  Hombre         Colombia  Sin Información  Sin Información   
184542  Hombre         Colombia  Sin Información  Sin Información   

            Escolaridad Pertenencia_grupal Pais_desaparicion        Municipio  \
1872    Sin información            Ninguno          Colombia  Sin Información   
2789    Sin información    Sin información          Colombia  Sin Información   
3750    Sin información    Sin información          Colombia  Sin Información   
3818    Sin información    Sin información          Colombia  Sin Información   
4310    Sin información    Sin información          Colombia  Sin Información   
...                 ...                ...               ...              ...   
151555  Sin información    Sin información          Colombia  Sin Información   
151740  Sin información    Sin información          Colombia  Sin Información   
174863  Sin información            Ninguno          Colombia  Sin Información   
175796  Sin información    Sin información          Colombia  Sin Información   
184542  Sin información    Sin información          Colombia  Sin Información   

           Departamento      Contexto  
1872    Sin Información  Desaparición  
2789    Sin Información  Desaparición  
3750    Sin Información  Desaparición  
3818    Sin Información  Desaparición  
4310    Sin Información  Desaparición  
...                 ...           ...  
151555  Sin Información  Desaparición  
151740  Sin Información  Desaparición  
174863  Sin Información  Desaparición  
175796  Sin Información  Desaparición  
184542  Sin Información  Desaparición  

[96 rows x 15 columns]

In [65]:
# Eliminar registros donde la mayoría de las columnas son 'Sin información'
umbral = 4
tabla_cleaned2 = tabla_cleaned[conteo_sin_info <= umbral]
print("Tabla después de eliminar registros con mayoría de 'Sin información':",len(tabla_cleaned2))
tabla_cleaned2


Tabla después de eliminar registros con mayoría de 'Sin información': 180155


ID         Estado                       Clasificacion  \
0            1   Desaparecido                     Sin Información   
1            2   Desaparecido                     Sin Información   
2            3   Desaparecido                     Sin Información   
3            4  Aparecio Vivo                     Sin Información   
5            6   Desaparecido                     Sin Información   
...        ...            ...                                 ...   
185736  185996   Desaparecido  Desaparición Presuntamente Forzada   
185737  185997   Desaparecido                     Sin Información   
185738  185998   Desaparecido                     Sin Información   
185739  185999   Desaparecido  Desaparición Presuntamente Forzada   
185740  186000   Desaparecido                     Sin Información   

             Fecha_desaparición  Edad    Sexo  Pais_nacimiento  \
0       2006-12-29T00:00:00.000  21.0  Hombre  Sin información   
1       2006-12-26T00:00:00.000  24.0  Hombre         Colombia   
2       2006-12-29T00:00:00.000  13.0   Mujer         Colombia   
3       2006-12-29T00:00:00.000  17.0   Mujer  Sin información   
5       2007-01-01T00:00:00.000  20.0   Mujer         Colombia   
...                         ...   ...     ...              ...   
185736  2002-11-10T00:00:00.000  39.0  Hombre         Colombia   
185737  2006-08-20T00:00:00.000  22.0  Hombre         Colombia   
185738  1992-04-01T00:00:00.000  18.0  Hombre         Colombia   
185739  2003-01-01T00:00:00.000  22.0  Hombre         Colombia   
185740  2012-11-14T00:00:00.000  24.0  Hombre         Colombia   

                   Raza     Estado_civil  \
0               Mestizo      Soltero (a)   
1               Mestizo      Soltero (a)   
2               Mestizo      Soltero (a)   
3               Mestizo      Soltero (a)   
5               Mestizo      Soltero (a)   
...                 ...              ...   
185736            Negro      Soltero (a)   
185737         Indigena      Soltero (a)   
185738  Sin Información      Soltero (a)   
185739  Sin Información      Unión Libre   
185740          Mestizo  Sin Información   

                                          Escolaridad  \
0       Educación básica secundaria o secundaria baja   
1       Educación básica secundaria o secundaria baja   
2       Educación básica secundaria o secundaria baja   
3         Educación técnica profesional y tecnológica   
5                           Educación básica primaria   
...                                               ...   
185736                                Sin escolaridad   
185737                                Sin información   
185738                                Sin información   
185739                      Educación básica primaria   
185740                                Sin información   

                                    Pertenencia_grupal Pais_desaparicion  \
0       Persona adicta a una droga natural o sintética          Colombia   
1       Persona adicta a una droga natural o sintética          Colombia   
2                                              Ninguno          Colombia   
3                                              Ninguno          Colombia   
5                                              Ninguno          Colombia   
...                                                ...               ...   
185736                                  Grupos étnicos          Colombia   
185737                                         Ninguno          Colombia   
185738                                         Ninguno          Colombia   
185739                                         Ninguno          Colombia   
185740                                            Otro          Colombia   

          Municipio Departamento      Contexto  
0       Bogotá D.C.  Bogotá D.C.  Desaparición  
1       Bogotá D.C.  Bogotá D.C.  Desaparición  
2       Bogotá D.C.  Bogotá D.C.  Desaparición  
3       Bogotá D.C.  Bogotá D.C.  Desaparición  
5       Bogotá D.C

**Nuestra tabla final para trabajar seria tabla_cleaned2 con un total de registros: 180155**